apps/prairielearn/src/pages/instructorInstanceAdminSettings/components/SelfEnrollmentSettings.tsx

```diff
@@ -86,6 +104,190 @@ export function SelfEnrollmentSettings({
+          name="show_in_enroll_page"
+        />
+        <label class="form-check-label" for="show_in_enroll_page">
+          Show on enrollment page
+        </label>
+        {showInEnrollPageError ? (
+          <div class="invalid-feedback">{showInEnrollPageError.message}</div>
+        ) : (
+          <div class="small text-muted">
+            If not checked, students will need a direct link to the course instance to enroll.
+          </div>
+        )}
+      </div>
+
+      <div
+        class={clsx(
+          'mb-3 form-check',
+          (!enrollmentManagementEnabled || !selfEnrollmentEnabled) && 'd-none',
+        )}
+      >
+        <input
+          class={clsx('form-check-input')}
+          type="checkbox"
+          id="self_enrollment_use_enrollment_code"
+          disabled={!canEdit || !selfEnrollmentEnabled}
+          {...control.register('self_enrollment_use_enrollment_code')}
+          name="self_enrollment_use_enrollment_code"
+        />
+        <label class="form-check-label" for="self_enrollment_use_enrollment_code">
+          Use enrollment code for self-enrollment
+        </label>
+        <div class="small text-muted">
+          If not checked, any link to the course instance will allow self-enrollment.
+        </div>
+      </div>
+
+      <div
+        class={clsx(
+          'mb-3 form-check',
+          (!enrollmentManagementEnabled || !selfEnrollmentEnabled) && 'd-none',
+        )}
+      >
+        <input
+          class={clsx('form-check-input')}
+          type="checkbox"
+          id="disable_self_enrollment_after_date"
+          disabled={!canEdit}
+          {...control.register('self_enrollment_enabled_before_date_enabled', {
+            onChange: async (event) => {
+              if (!event.target.checked) {
+                await trigger('self_enrollment_enabled_before_date');
+              }
+            },
+          })}
+          name="self_enrollment_enabled_before_date_enabled"
+        />
+        <label class="form-check-label" for="disable_self_enrollment_after_date">
+          Forbid self-enrollment after specified date
+        </label>
+
+        <input
+          type="datetime-local"
+          aria-label="Self-enrollment enabled before date"
+          class={clsx('form-control mt-2', selfEnrollmentEnabledBeforeDateInvalid && 'is-invalid')}
+          disabled={!canEdit || !selfEnrollmentEnabledBeforeDateEnabled}
+          {...control.register('self_enrollment_enabled_before_date', {
+            validate: (value, { self_enrollment_enabled_before_date_enabled }) => {
+              if (self_enrollment_enabled_before_date_enabled && !value) {
+                return 'Date is required';
+              }
+              return true;
+            },
+          })}
+          // HACK: Make the accessibility checker happy
+          {...(!selfEnrollmentEnabledBeforeDateEnabled
+            ? { name: 'self_enrollment_enabled_before_date__IGNORE' }
+            : {})}
```

Comment:
There's a relevant issue in `html-validator` right? Consider linking it here and spelling out why it's unhappy in the first place?